core/src/main/java/com/google/cloud/sql/core/CloudSqlInstance.java

```diff
@@ -0,0 +1,496 @@
+              connectionName));
+    }
+
+    // Parse the certificate from the response.
+    Certificate ephemeralCertificate;
+    try {
+      byte[] certBytes = response.getCert().getBytes(StandardCharsets.UTF_8);
+      ephemeralCertificate =
+          CertificateFactory.getInstance("X.509")
+              .generateCertificate(new ByteArrayInputStream(certBytes));
+    } catch (CertificateException ex) {
+      throw new RuntimeException(
+          String.format(
+              "[%s] Unable to parse the ephemeral certificate for the Cloud SQL instance.",
+              connectionName),
+          ex);
+    }
+    return ephemeralCertificate;
+  }
+
+  // Schedules task to be executed once the provided futures are complete.
+  private <T> ListenableFuture<T> whenComplete(Callable<T> task, ListenableFuture<?>... futures) {
+    SettableFuture<T> taskFuture = SettableFuture.create();
+
+    // Create a countDown for all Futures to complete.
+    AtomicInteger countDown = new AtomicInteger(futures.length);
+
+    // Trigger the task when all futures are complete.
+    Runnable runOutput =
+        () -> {
+          if (countDown.decrementAndGet() == 0) {
+            taskFuture.setFuture(executor.submit(task));
+          }
+        };
+    for (ListenableFuture<?> future : futures) {
+      future.addListener(runOutput, executor);
+    }
+
+    return taskFuture;
+  }
+
+  // Returns the inner future from a nested future, or else the exception it encountered.
+  private <T> ListenableFuture<T> extractNestedFuture(
+      ListenableFuture<ListenableFuture<T>> future) {
+    try {
+      return Futures.getDone(future);
+    } catch (ExecutionException ex) {
+      return Futures.immediateFailedFuture(ex.getCause());
+    }
+  }
+
+  /**
+   * Checks for common errors that can occur when interacting with the Cloud SQL Admin API, and adds
+   * additional context to help the user troubleshoot them.
+   *
+   * @param ex The exception thrown by the Admin API request.
+   * @param fallbackDesc Generic description uses as a fall back if no additional information can be
+   *     provided to the user.
+   */
+  // TODO(kvg) Throw checked exceptions for recoverable states?
+  private RuntimeException addExceptionContext(IOException ex, String fallbackDesc) {
+    // Verify we are able to extract a reason from an exception, or fallback to a generic desc
+    GoogleJsonResponseException gjrEx =
+        ex instanceof GoogleJsonResponseException ? (GoogleJsonResponseException) ex : null;
+    if (gjrEx == null
+        || gjrEx.getDetails() == null
+        || gjrEx.getDetails().getErrors() == null
+        || gjrEx.getDetails().getErrors().isEmpty()) {
+      return new RuntimeException(fallbackDesc, ex);
+    }
+    // Check for commonly occurring user errors and add additional context
+    String reason = gjrEx.getDetails().getErrors().get(0).getReason();
+    if ("accessNotConfigured".equals(reason)) {
```

Comment:
Up to you, but you can use `switch` with string literal cases.

Reply:
I didn't use it here because I'm not certain reason won't be null safe.